wallet-new/integration/QuickCheckSpecs.hs

```diff
@@ -0,0 +1,115 @@
+        it "V0 API follows best practices & is RESTful abiding" $ do
+            serverSatisfies (Proxy @V0.API) burl stdArgs predicates
+        it "V1 API follows best practices & is RESTful abiding" $ do
+            serverSatisfies (Proxy @V1.API) burl stdArgs predicates
+--
+-- Instances to allow use of `servant-quickcheck`.
+--
+
+instance HasGenRequest (apiType a :> sub) =>
+         HasGenRequest (WithDefaultApiArg apiType a :> sub) where
+    genRequest _ = genRequest (Proxy @(apiType a :> sub))
+
+instance HasGenRequest (argA a :> argB a :> sub) =>
+         HasGenRequest (AlternativeApiArg argA argB a :> sub) where
+    genRequest _ = genRequest (Proxy @(argA a :> argB a :> sub))
+
+-- NOTE(adinapoli): This can be improved to produce proper filtering & sorting
+-- queries.
+instance HasGenRequest sub => HasGenRequest (SortBy syms res :> sub) where
+    genRequest _ = genRequest (Proxy @sub)
+
+instance HasGenRequest sub => HasGenRequest (FilterBy syms res :> sub) where
+    genRequest _ = genRequest (Proxy @sub)
+
+instance HasGenRequest sub => HasGenRequest (Tags tags :> sub) where
+    genRequest _ = genRequest (Proxy :: Proxy sub)
+
+instance HasGenRequest (sub :: *) => HasGenRequest (WalletRequestParams :> sub) where
+    genRequest _ = genRequest (Proxy @(WithWalletRequestParams sub))
+
+--
+-- RESTful-abiding predicates
+--
+
+-- | Checks that every DELETE request should return a 204 NoContent.
+deleteReqShouldReturn204 :: RequestPredicate
+deleteReqShouldReturn204 = RequestPredicate $ \req mgr ->
+    if (method req == methodDelete)
+        then do
+            resp <- httpLbs req mgr
+            let status = responseStatus resp
+            when (statusIsSuccessful status && status /= status204) $
+                throwM $ PredicateFailure "deleteReqShouldReturn204" (Just req) resp
+            return [resp]
+        else
+            return []
+
+-- | Checks that every PUT request is idempotent. Calling an endpoint with a PUT
+-- twice should return the same result.
+putIdempotency :: RequestPredicate
+putIdempotency = RequestPredicate $ \req mgr ->
+     if (method req == methodPut)
+         then do
+             resp1 <- httpLbs req mgr
+             resp2 <- httpLbs req mgr
+             let body1 = responseBody resp1
+             let body2 = responseBody resp2
+             when (body1 /= body2) $
+                 throwM $ PredicateFailure "putIdempotency" (Just req) resp1
+             return [resp1, resp2]
+         else
+             return []
+
+-- | Checks that every request which is not a 204 No Content
+-- does not have an empty body, but it always returns something.
+noEmptyBody :: RequestPredicate
+noEmptyBody = RequestPredicate $ \req mgr -> do
+    resp <- httpLbs req mgr
+    let body   = responseBody resp
+    let status = responseStatus resp
+    when (status `notElem` [status204, status404] && body == mempty) $
+        throwM $ PredicateFailure "noEmptyBody" (Just req) resp
```

Comment:
Why include `status404` here as well? Don't we return a body with a clear message for those :s ?